### ISSUE TITLE
Old leader might confirm data not present on new leader

### ISSUE THREAD

**Issue by sergepetrenko** [tarantool/tarantool#7253]
<details><summary>10 step reproducer:</summary>

Run 3 tarantools in 3 consoles and execute the following commands in respective consoles:
```lua
--
-- Instance 1
--

-- Step 1
box.cfg{listen = 3301, replication = {3301,3302,3303}, election_mode = 'candidate', election_fencing_enabled = false, replication_timeout = 10}

-- Step 4
box.schema.user.grant('guest', 'replication')
box.schema.space.create('sync', {is_sync = true})
box.space.sync:create_index('pk')

-- Step 5
box.cfg{replication = ""}

-- Step 9
box.space.sync:insert{1}

--
-- Instance 2
--

-- Step 2
box.cfg{listen=3302, replication={3301,3302,3303}, election_mode='voter', replication_timeout = 10}

--
-- Instance 3
--

-- Step 3
box.cfg{listen=3303, replication={3301,3302,3303}, election_mode='voter', replication_timeout = 10}

-- Step 6
box.cfg{replication = {3302, 3303}}

-- Step 7
box.cfg{election_mode = 'candidate'}

-- Step 8
box.ctl.promote() box.error.injection.set("ERRINJ_WAL_DELAY", true) -- important to do this in one line, whitout yields.

-- Step 10
box.error.injection.set("ERRINJ_WAL_DELAY", false)
```

</details>

After you've performed these 10 steps:
 * Instance 3 is leader
 * Instance 1 and instance 2 output:
```lua
tarantool> box.space.sync:get{1}
---
- [1]
...
```
 * Instance 3 output:
```lua
tarantool> box.space.sync:get{1}
---
...
```

I guess it's already obvious what happened, but still some excerpts from the node WALs for convenience.
<details><summary>Instance 1 xlog</summary>

```
---
HEADER:
  lsn: 9
  replica_id: 1
  type: INSERT
  timestamp: 1654768508.6078
BODY:
  space_id: 512
  tuple: [1]
---
HEADER:
  lsn: 10
  replica_id: 1
  type: CONFIRM
  timestamp: 1654768508.608
BODY:
- null
- 1
- 9
...
```

</details>
<details><summary>Instance 2 xlog</summary>

```
---
HEADER:
  lsn: 9
  replica_id: 1
  type: INSERT
  timestamp: 1654768508.6079
BODY:
  space_id: 512
  tuple: [1]
---
HEADER:
  lsn: 10
  replica_id: 1
  type: CONFIRM
  timestamp: 1654768508.6081
BODY:
- null
- 1
- 9
---
HEADER:
  lsn: 1
  replica_id: 2
  type: PROMOTE
  timestamp: 1654768510.2046
BODY:
  2: 1
  3: 0
  83: 4
...
```

</details>
<details><summary>Instance 3 xlog</summary>

```
---
HEADER:
  lsn: 1
  replica_id: 2
  type: PROMOTE
  timestamp: 1654768508.0623
BODY:
  2: 1
  3: 0
  83: 4
---
HEADER:
  lsn: 9
  replica_id: 1
  type: NOP  <-- this was INSERT
  timestamp: 1654768510.2045
---
HEADER:
  lsn: 10
  replica_id: 1
  type: NOP  <-- this was CONFIRM
  timestamp: 1654768510.2045
...
```

</details>

**Note:** it's not necessary to turn the fencing off on the old leader or disconnect it from everyone to reproduce the issue. It's just easier to reproduce with fencing = off. The same might happen with soft fencing, and even with strict fencing, when manual elections are in play.
**Note:** the issue may be reproduced without the manual elections. When fencing is not strict.

**Issue №1:** new leader doesn't ask the quorum of their actual vclock before writing PROMOTE.
**Issue №2:** even if new leader did so, old leader might always insert and confirm something new while new leader is stuck writing PROMOTE.
**Issue №3:** even if we disallow replication from everyone but the leader (like it's done in canonical Raft), it's still possible that old leader's data enters WAL write on a voter BEFORE vote request from the new leader arrives, but  WAL write confirmation arrives AFTER the voter votes for the new leader. This means that a voter votes for a leader which doesn't have the same data, as voter does.
**Issue №4:** when #5295 is fixed, it will detect this situation, and will force the cluster partition. Moreover, the new leader will be left in a smaller portion of the cluster. Essentially not able to commit anything.

**Proposed steps to fix:**
1. **On all nodes:** do not accept replication from ANY node, whose term is smaller, than ours. This is a compromise between what we had initially (applying replication only from the leader) and what we have now (applying replication from everyone, but nopifying it, when it comes from an old journal term).
2. **On old leader:** freeze the limbo not only during fencing, but always when stepping off. Due to new term, for example. So that the old leader doesn't confirm / rollback some transactions which should be overtaken by the new leader. This will minimise chances of a split-brain.
3. **On followers:** when voting for a candidate, take into account not only `replicaset.vclock`, but `replicaset.applier.vclock`, because the latter might reflect changes that came from an old leader and are now undergoing a WAL write.

**Comment by Gerold103** [comment]
> Issue №1: new leader doesn't ask the quorum of their actual vclock before writing PROMOTE

Why is it possible? Quorum wouldn't vote for the new leader if their vclock is bigger or incomparable.

> Issue №3: 

`wal_sync()` before responding to a vote request could help?

**Comment by sergepetrenko** [comment]
> Why is it possible? Quorum wouldn't vote for the new leader if their vclock is bigger or incomparable.

That's because their vclock might increase after they vote for the new leader.
So by the time the leader writes a PROMOTE, they have some new data. And this means the leader fails to confirm some data written by the old leader. He rolls it back.

> wal_sync() before responding to a vote request could help?

It seems so

**Comment by Gerold103** [comment]
I did some initial investigation. While I think what to do with that, I post it all here. And will most likely update this post in-place with more info later.

#### The avoidable split brain

On the latest Tarantool node1 and node2 are in fact in the majority of the new term. The new leader was cut off due to split brain. Also I can't see any data inconsistency. Both old leader and the middle node have matching terms, good connectivity, and `{1}` tuple. I would even say all is fine, but the issue is still that
- The tuple `{1}` was applied on node2 when it already saw the new term 3. It just didn't get a new `PROMOTE` yet.
- This entire split-brain could be avoided. Below I try to think about how.

With this tuple `{1}` the natural solution to me seems to ask node1 somehow to wait. Say to it like "hey, your term is old, freeze your limbo and wait until we elect a new leader in the current term".

Problem is that node1 doesn't even see there is a new term. Because we send raft messages to relays, not to appliers. Hence node1 can't see there is a new term.

If we do that, the node1 will see there is a new term, will freeze the limbo, and wait for term outcome. Node2 applies `{1}`, but node1 won't emit `CONFIRM` nor `ROLLBACK` - both would lead to split-brain. This `{1}` txn now can only be finished with a `PROMOTE`. Rolled back or committed - depends on what the new leader will decide.

Why can't node2 nopify `{1}`? The reasons are:
- Until the term outcome is known, we would have node1 and node2, both connected just fine, but having different dataset. That is inconsistent.
- Node1 actually might gather majority from some other nodes not visible to node2 and node3. So it will later emit a rightful `CONFIRM` and what would node2 do now?

Why can't node2 reject node1 with split-brain error when it gets `{1}` in the old term? Because it would only increase split-brain probability.

#### Vote while having pending WAL writes

That seems to be a separate issue, not related to the situation above. Indeed, one solution could be to look at `replicaset.applier.vclock`. But
- This one can go backwards if txn fails to commit. Doesn't seem reliable to use it anywhere except how appliers use it.
- It would mean we either need to switch the entire Raft to it, or have 2 vclocks in Raft. I am very reluctant about growing Raft state if it can be avoided.
- `Replicaset.applier.vclock` won't work if the old leader will start a WAL write. It won't affect `replicaset.applier.vclock`, but will affect the main vclock after WAL write is done.

The solution would be to call `wal_sync()` before voting. But it would complicate Raft further - we would need to add a new method to `raft_vtab` and a new async job in `raft_process_async()`.

But it seems we could also split term and vote WAL writes into separate xrows. Then before persisting `volatile_vote` we again check if it can be done. If it can't be done, then we cancel the vote.

For that we would need to save the vote vclock into struct raft. So as we could check it after term WAL write.

Why not do WAL sync right in `raft_process_msg()`? Because this function is not supposed to yield. If it would yield, we would need to re-validate all the checks done before the WAL sync because the term could be bumped again, leader could be already found, raft could be disabled, and so on. While it doesn't yield it can be kept simple.

The potential problem affecting both `wal_sync()` way and split term and vote WAL writes way, is that during `wal_sync()` or term WAL write new transactions could be sent to WAL. Perhaps it can only happen for async transactions, because for the sync ones the limbo would be frozen as there is a new term but no leader yet. That would need to be checked. Don't know what will happen if after all a vote is given to a node with older vclock.

The solution to new WAL writes after syncing WAL but before writing vote would be to somehow lock new WAL writes. The lock duration would be short - just one round-trip to WAL thread for the sync itself and one WAL write of the vote.

**Comment by kostja** [comment]
> Problem is that node1 doesn't even see there is a new term. Because we send raft messages to relays, not to appliers.

What can I say... I told you 4 years ago applier and relay should move to WAL thread and stay together like in most raft implementations. You did everything to avoid it. Please stop re-inventing the wheel and do raft by the spec.

**Comment by Gerold103** [comment]
I don't see how it is related. If they would be in WAL, we still would send it only to relays and would have exactly the same problem. To appliers it wasn't sent I don't remember why tbh. Probably we didn't want to complicate applier's protocol too, but probably it is time. Also please, stop posting exposing your hard feelings about this WAL-fixation everywhere you can reach.

As for the raft-by-the-spec - please, be more specific what is not according to the spec in tnt which leads to the problem described in the ticket. If I didn't mention it in my comment above already. And how you suggest to fix it.

**Comment by kostja** [comment]
And of course ironically when you try to follow the standard raft you'll find it very difficult to do if the logic of the protocol is split between different threads.

**Comment by kostja** [comment]
> > if you don't have a majority of acks, and if you do have it, this data must be on the new leader
> 
> Correct. Here we just have a bug it seems. That the node2 manages to ack a txn even knowing there is a bigger term going.
> 
> > The new leader always appends a dummy entry after election to commit all entries from the previous terms before new entries are committed
> 
> We do that with `PROMOTE` record. The problem here happens before the election ends. Node3 is writing this dummy entry/`PROMOTE`, but until it is done, node2 keeps acking txns from the previous term. So the solution should somehow make node2 not ack that outdated txn but not reject it as an inconsistency either, because node1 will roll it back after gets `PROMOTE` from node3 in the newest term.
> 
> > But since you moved configuration changes outside Raft
> 
> Reconfiguration of the cluster has absolutely nothing to do with that. The test doesn't involve it at all.
> 
> > And of course ironically when you try to follow the standard raft you'll find it very difficult to do if the logic of the protocol is split between different threads.
> 
> The entire raft logic is in TX thread. Applier and relay only do socket work and heartbeats. Even if you would have them together in one thread, WAL or not, it wouldn't be the TX thread anyway, because we can't afford blocking TX for socket IO. And it wouldn't help regardless. Sending records to a thread which will make IO has the same guarantees as doing the IO straight away - unknown when delivered, unknown whether delivered at all, unknown whether there is a race with another node doing same IO and so on.

It's very hard to understand your reply due to terminology differences, but I will try. You write that the newly elected leader "keeps acking" entries from the old term. This violates the basic Raft rule - any kind of AppendEntries with an older term than yours gets rejected. If you look at etcd raft or other raft implementations, there is a common chunk of code in step() function which would address the common protocol rules without even looking what's inside the patcket, and this is one of them. Ignore messages iwth outdated terms.

**Comment by Gerold103** [comment]
A reproducer for the bug when node during voting doesn't account txns whose WAL write is in progress.
```Lua
--
-- Instance 1
--
-- Step 1
--
box.cfg{
    listen = 3313,
    election_mode = 'manual',
    replication = {3313, 3314, 3315},
    replication_synchro_timeout = 1000,
    replication_synchro_quorum = 2,
}
box.schema.user.grant('guest', 'super')

-- Step 4
--
s1 = box.schema.create_space('test1', {is_sync = true})
_ = s1:create_index('pk')

-- Step 5
--
box.cfg{replication = {3313, 3314}}

-- Step 8
--
fiber = require('fiber')
f = fiber.create(s1.replace, s1, {1})


--
-- Instance 2
--
-- Step 2
--
box.cfg{
    listen = 3314,
    election_mode = 'voter',
    replication = {3313, 3314, 3315},
    replication_synchro_timeout = 1000,
    replication_synchro_quorum = 2,
}

-- Step 7
--
box.error.injection.set('ERRINJ_WAL_DELAY', true)

-- Step 10
--
box.error.injection.set('ERRINJ_WAL_DELAY', false)


--
-- Instance 3
--
-- Step 3
--
box.cfg{
    listen = 3315,
    election_mode = 'voter',
    election_timeout = 1000,
    replication = {3313, 3314, 3315},
    replication_synchro_timeout = 1000,
    replication_synchro_quorum = 2,
}

-- Step 6
--
box.cfg{replication = {3314, 3315}}

-- Step 9
--
box.cfg{election_mode = 'candidate'}
box.ctl.promote()
```
Behaviour is different from time to time. One of the failures - node2 and node3 stay connected bidirectionally. Both have `{1}` tuple. Node1 knows about the new term, but doesn't know the leader (because we only set it when receive anything from it directly) and still has downstream from node2.
The issues:
- Node1 still has this `{1}` transaction not finished. The fiber keep hanging. At the same time its limbo len is 0. And changing `replication_synchro_timeout` doesn't do anything. Looks like the transaction got deleted from the limbo without an outcome.
- Node2 shouldn't have voted for node3 when it knew about `{1}` txn ongoing. It would prevent this split-brain case.

**Comment by kostja** [comment]
There are no options in Raft. Whenever you get an entry with a bigger term than yours, you *always* update your term, no matter what kind of entry this is. The exception is pre-vote, which is having a term in the future, but you don't have the pre-vote packets anyway. 

When you get an entry with an older term, you reply AppendEntries::Reject message, notifying the sender you don't accept that entry. The rejects are not counted in quorum.

**Comment by kostja** [comment]
When a node steps down  it becomes a follower. So it should stop streaming data right away. All its pending write queue should be aborted. If these transactions later come from the new leader, they will be appended to the log with a new term.

**Comment by Gerold103** [comment]
> Whenever you get an entry with a bigger term than yours, you always update your term, no matter what kind of entry this is

Node2 saw new term and applied it just fine. Node1 didn't see the new term yet so it wasn't expected to bump it locally. It is not the problem.

> When you get an entry with an older term, you reply with AppendEntries::Reject message, notifying the sender you don't accept that entry.

Yes, we don't have this 'reject' packet. It seems to be the problem. Node2 should notify node1 that there is a newer term by explicitly rejecting this txn from the old term. I explained it in my [comment](https://github.com/tarantool/tarantool/issues/7253#issuecomment-1183680715) above as "third option".

> All its pending write queue should be aborted

If we rollback the txns and write `ROLLBACK` into the journal, it might trigger `ER_SPLIT_BRAIN`. Because new leader will receive this `ROLLBACK` from an older term eventually and rightfully will suspect that the old leader is still making decisions about synchro txns. Moreover, you can't really rollback it if just one node responded with a reject. All the others could get the txn and not respond anything due to a network issue. The new leader could also already get your data. Then it could collect quorum and commit your record. So you can't really abort the queue until you have more info. Raft AFAIR doesn't do that either. In this particular case, when one instance rejected the entry, you can't decide anything yet.

**Comment by Gerold103** [comment]
Reproducer for the hanging fiber when its txn is already finished. Isolated from the other bugs.
```Lua
--
-- Instance 1
--
-- Step 1
--
box.cfg{
    listen = 3313,
    election_mode = 'manual',
    replication = {3313, 3314, 3315},
    replication_synchro_timeout = 1000,
    replication_synchro_quorum = 2,
}
box.schema.user.grant('guest', 'super')

-- Step 4
--
s1 = box.schema.create_space('test1', {is_sync = true})
_ = s1:create_index('pk')

-- Step 8
--
fiber = require('fiber')
f = fiber.create(s1.replace, s1, {1})

-- Step 9
--
box.error.injection.set('ERRINJ_WAL_DELAY', true)

-- Step 11
-- Wait confirm being written.

-- Step 13
--
box.error.injection.set('ERRINJ_WAL_DELAY', false)


--
-- Instance 2
--
-- Step 2
--
box.cfg{
    listen = 3314,
    election_mode = 'voter',
    replication = {3313, 3314, 3315},
    replication_synchro_timeout = 1000,
    replication_synchro_quorum = 2,
}

-- Step 5
--
box.cfg{replication = {3313, 3314}}

-- Step 7
--
box.error.injection.set('ERRINJ_WAL_DELAY', true)

-- Step 10
--
box.error.injection.set('ERRINJ_WAL_DELAY', false)


--
-- Instance 3
--
-- Step 3
--
box.cfg{
    listen = 3315,
    election_mode = 'voter',
    election_timeout = 1000,
    replication = {3313, 3314, 3315},
    replication_synchro_timeout = 1000,
    replication_synchro_quorum = 2,
}

-- Step 6
--
box.cfg{replication = {}}

-- Step 12
--
box.cfg{election_mode = 'candidate'}
box.ctl.promote()
```
The test achieves that the limbo fencing on the first instance is activated during `CONFIRM` WAL write. Limbo entry goes to waiting without checking if it is already completed. After step 13 can see that `f` is `suspended`. But the limbo queue len is 0. The fiber finishes when you just do `f:wakeup()`.

A trivial fix would be to check if the limbo entry is completed before going to the waiting, but beforehand need to think twice if this fencing during `CONFIRM` is legal at all.

**Comment by sergepetrenko** [comment]
> A trivial fix would be to check if the limbo entry is completed before going to the waiting, but beforehand need to think twice if this fencing during CONFIRM is legal at all.

I suppose all the CONFIRMs and ROLLBACKs that are already sent to WAL at the moment of fencing should be processed as usual.
